crates/vm/src/system/public_values/core.rs

```diff
@@ -169,32 +170,48 @@ where
         trace_offset: &mut usize,
         width: usize,
     ) -> Result<()> {
-        todo!("Implement execute function");
-    }
+        let row_slice = &mut trace[*trace_offset..*trace_offset + width];
+        let (adapter_row, core_row) = unsafe { row_slice.split_at_mut_unchecked(A::WIDTH) };
 
-    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row_slice: &mut [F]) {
-        todo!("Implement fill_trace_row function");
+        let [[value], [index]] = self.adapter.read(state.memory, instruction, adapter_row);
+        {
+            let idx: usize = index.as_canonical_u32() as usize;
+            let mut custom_pvs = self.custom_pvs.lock().unwrap();
 
-        // let (adapter_row, core_row) = unsafe { row_slice.split_at_mut_unchecked(A::WIDTH) };
+            if custom_pvs[idx].is_none() {
+                custom_pvs[idx] = Some(value);
+            } else {
+                // Not a hard constraint violation when publishing the same value twice but the
+                // program should avoid that.
+                panic!("Custom public value {} already set", idx);
+            }
+        }
 
-        // self.adapter.fill_trace_row(mem_helper, (), adapter_row);
+        let cols = PublicValuesCoreColsView::<_, &mut F>::borrow_mut(core_row);
+        debug_assert_eq!(cols.width(), width - A::WIDTH);
 
-        // let core_row: &mut PublicValuesCoreColsView<_, F> = core_row.borrow_mut();
+        *cols.is_valid = F::ONE;
+        *cols.value = value;
+        *cols.index = index;
+
+        *state.pc = state.pc.wrapping_add(DEFAULT_PC_STEP);
+        *trace_offset += width;
 
-        // // TODO(ayush): add this check
-        // // debug_assert_eq!(core_row.width(), BaseAir::<F>::width(&self.air));
+        Ok(())
+    }
 
-        // core_row.is_valid = F::ONE;
-        // core_row.value = record.value;
-        // core_row.index = record.index;
+    fn fill_trace_row(&self, mem_helper: &MemoryAuxColsFactory<F>, row_slice: &mut [F]) {
+        let (adapter_row, core_row) = unsafe { row_slice.split_at_mut_unchecked(A::WIDTH) };
```

Comment:
```suggestion
        let (adapter_row, core_row) = row_slice.split_at_mut(A::WIDTH);
```

Is there a benefit to doing an unsafe split?

Reply:
i think it avoids bounds checking at runtime. not sure if the marginal benefit is worth it though. i just followed the patter of how it was being done in other areas of the code